src/serenity/config.hpp

```diff
@@ -1,90 +1,222 @@
+ */
+using CfgVariant =
+  boost::variant<bool, int64_t, uint64_t, double_t, std::string>;
+
 
 /**
- * State for every Change Point Detector. It enables user to
- * configure change Point Detectors.
- * Detectors are also able to adjust these values dynamically.
+ * Global Serenity Config class which implements basic mechanism
+ * to support specifying config parameters via string key map.
+ *
+ * Check config_test.cpp to see example usage.
  */
-struct ChangePointDetectionState {
-  static ChangePointDetectionState createForNaiveDetector(
-      uint64_t _contentionCooldown,
-      double_t _absoluteThreshold) {
-    ChangePointDetectionState state;
-    state.contentionCooldown = _contentionCooldown;
-    state.absoluteThreshold = _absoluteThreshold;
-
-    return state;
-  }
-
-  static ChangePointDetectionState createForRollingDetector(
-      uint64_t _windowSize,
-      uint64_t _contentionCooldown,
-      double_t _relativeThreshold) {
-    ChangePointDetectionState state;
-    state.windowSize = _windowSize;
-    state.contentionCooldown = _contentionCooldown;
-    state.relativeThreshold = _relativeThreshold;
-
-    return state;
-  }
-
-  uint64_t windowSize = changepoint::DEFAULT_WINDOW_SIZE;
-  //! How many iterations detector will wait with creating another
-  //! contention.
-  uint64_t contentionCooldown = changepoint::DEFAULT_CONTENTION_COOLDOWN;
-
-  //! NaiveChangePointDetector bases its filtering on absolute value.
-  //! Below that value detector will trigger contention.
-  double_t absoluteThreshold = changepoint::DEFAULT_ABS_THRESHOLD;
-
-  //! Defines how much value must drop to trigger contention.
-  //! Most detectors will use that.
-  double_t relativeThreshold = changepoint::DEFAULT_RELATIVE_THRESHOLD;
-
-  //! Defines how much (relatively to base point) value must drop to trigger
-  //! contention.
-  //! Most detectors will use that.
-  double_t fractionalThreshold =
-    changepoint::DEFAULT_FRACTIONAL_THRESHOLD;
-
-  //! You can adjust how big severity is created for  a defined drop.
-  double_t severityLevel = changepoint::DEFAULT_SEVERITY_FRACTION;
-
-  //! Tolerance fraction of threshold if signal is accepted as returned to
-  //! previous state after drop.
-  double_t nearFraction = changepoint::DEFAULT_NEAR_FRACTION;
-};
+class SerenityConfig {
+ public:
+  SerenityConfig() {}
+
+  /**
+   * Overlapping custom configuration options using recursive copy.
+   */
+  void cpy(const SerenityConfig& customCfg) {
+    this->recursiveCfgCopy(this, customCfg);
+  }
+
+  /**
+   * Recursive copy.
+   */
+  void recursiveCfgCopy(SerenityConfig* base,
+                        const SerenityConfig& customCfg) const {
+    for (auto customItem : customCfg.fields) {
+      base->fields[customItem.first] = customItem.second;
+    }
+
+    for (auto customSection : customCfg.sections) {
+      this->recursiveCfgCopy(
+        &((*base)[customSection.first]), *customSection.second);
```

Comment:
/serenity/src/serenity/config.hpp:45:  Is this a non-const reference? If so, make const or use a pointer: SerenityConfig& base  [runtime/references] [2]